pkg/agent/proxy/types/types.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/vmware-tanzu/antrea/pkg/ovs/openflow"
+	k8sproxy "github.com/vmware-tanzu/antrea/third_party/proxy"
+)
+
+// ServiceInfo is the internal struct for string service information
+type ServiceInfo struct {
+	*k8sproxy.BaseServiceInfo
+	// cache for performance
+	OFTransportProtocol    openflow.Protocol
+	AffinityTimeoutSeconds uint16
+}
+
+func (si *ServiceInfo) Equal(bSvcInfo *ServiceInfo) bool {
+	return si.SessionAffinityType() == bSvcInfo.SessionAffinityType() &&
+		si.StickyMaxAgeSeconds() == bSvcInfo.StickyMaxAgeSeconds() &&
+		si.OFTransportProtocol == bSvcInfo.OFTransportProtocol &&
+		si.Port() == bSvcInfo.Port()
+}
+
+// NewServiceInfo returns a new upstream.ServicePort which abstracts a serviceInfo
+func NewServiceInfo(port *corev1.ServicePort, service *corev1.Service, baseInfo *k8sproxy.BaseServiceInfo) k8sproxy.ServicePort {
+	info := &ServiceInfo{BaseServiceInfo: baseInfo}
+	info.OFTransportProtocol = openflow.ProtocolTCP
+	if port.Protocol == corev1.ProtocolUDP {
+		info.OFTransportProtocol = openflow.ProtocolUDP
+	} else if port.Protocol == corev1.ProtocolSCTP {
+		info.OFTransportProtocol = openflow.ProtocolSCTP
+	}
+	if info.SessionAffinityType() == corev1.ServiceAffinityClientIP {
+		info.AffinityTimeoutSeconds = uint16(info.StickyMaxAgeSeconds())
+		if info.AffinityTimeoutSeconds == 0 {
+			info.AffinityTimeoutSeconds = uint16(corev1.DefaultClientIPServiceAffinitySeconds)
```

Comment:
I think newBaseServiceInfo has taken care of this and you should use `info.StickyMaxAgeSeconds()` only? and no need to have another redundant field?
https://github.com/vmware-tanzu/antrea/blob/e9062a9383c796d7b17e792de40116a3af08172f/third_party/proxy/service.go#L151-L153

Reply:
The timeout value may be 0 and I need to set it as the default value. Since the timeout field in upstream code is not exportable, I would like to have a add-on field here.

Reply:
I think K8s has validated that the timeout must not be 0, otherwise kube-proxy code would have done the defaulting.

Reply:
Just checked upstream code, in pkg/apis/core/v1/defaults.go:104, the default value will be set if it is absented.
Updated.